fix(transformer/jsx): fix parsing JSX pragma comments

overlookmotel opened the pull request:
Fixes #10955.

Allow multiple JSX pragmas in a single comment, or even in the same line. e.g.

```js
/* @jsx h @jsxRuntime classic */

// @jsx h @jsxRuntime classic

/**
  * @jsx h
  * @jsxRuntime classic
  */
```

Also, allow pragmas to be anywhere in the comment, including surrounded by unrelated text.

```js
// Unrelated text @jsx h Unrelated text @jsxRuntime classic

/**
  * Unrelated text.
  * @jsx h
  * Unrelated text.
  * @jsxRuntime classic
  */
```

This is more liberal than Babel, which [has a test](https://github.com/babel/babel/blob/226318d54eaff275500b2723c46ee608c0632df0/packages/babel-plugin-transform-react-jsx/test/fixtures/react/should-not-allow-jsx-pragma-to-be-anywhere-in-comment/input.js) specifically to outlaw placing pragmas amongst other text in a comment - this test is disabled in this PR.

Instead, we're aligning with ESBuild ([ESBuild playground](https://esbuild.github.io/try/#dAAwLjI1LjQALS1sb2FkZXI9anN4IC0tanN4PWF1dG9tYXRpYwAvLyBAanN4IFNvbWV0aGluZyBibGFoIGJsYWggQGpzeFJ1bnRpbWUgY2xhc3NpYwo8YmxhaC8+Cg)) which seems to allow a `@jsx` pragma anywhere in the text of the comment.

overlookmotel commented:
* **#10983** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/10983?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/10983?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add either label to this PR to merge it via the merge queue:
- <kbd>0-merge</kbd> - adds this PR to the back of the merge queue
- <kbd>hotfix</kbd> - for urgent hot fixes, skip the queue and merge this PR next

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.dev/invite/github/oxc-project?ref=merge-queue-instructions-in-stack-comment&prId=10983).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

overlookmotel commented:
> ❌	`transformer[antd.js]`	40.8 ms	42.3 ms	-3.53%

I'll try to address this perf regression in a follow-up.

overlookmotel commented:
@Dunqing Do you think we should be less liberal than ESBuild? Accepting the `@jsx` pragma in this comment seems unwise.

```js
// Don't use pragma here @jsx React.createDumpling
```

We could instead accept multiple pragmas in the same comment, or on the same line, but outlaw unrelated text around them.

i.e. Make these acceptable:

```js
/* @jsx h @jsxRuntime classic */

/**
  * @jsx h
  * @jsxRuntime classic
  */
```

But these are not:

```js
// Unrelated text @jsx h Unrelated text @jsxRuntime classic

/**
  * Unrelated text.
  * @jsx h
  * Unrelated text.
  * @jsxRuntime classic
  */
```

I wonder if ESBuild did it the maximally liberal way just because it's the easiest to implement, not because it's the desired behavior.

Dunqing commented:
> > ❌	`transformer[antd.js]`	40.8 ms	42.3 ms	-3.53%
> 
> I'll try to address this perf regression in a follow-up.

Weird, shouldn't have any performance impact in this benchmark, as it applies when the source type is JSX. Could you investigate why this is happening?
https://github.com/oxc-project/oxc/blob/287b3c368f5d7a1fafcc85a6532ed72606627fdb/crates/oxc_transformer/src/lib.rs#L123-L130

Dunqing commented:
### Merge activity

* **May 12, 8:28 PM EDT**: The merge label '0-merge' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.dev/merges?org=oxc-project&repo=oxc) once it meets the requirements.

overlookmotel committed:
fix(transformer/jsx): fix parsing JSX pragma comments (#10983)

Fixes #10955.

Allow multiple JSX pragmas in a single comment, or even in the same line. e.g.

```js
/* @jsx h @jsxRuntime classic */

// @jsx h @jsxRuntime classic

/**
  * @jsx h
  * @jsxRuntime classic
  */
```

Also, allow pragmas to be anywhere in the comment, including surrounded by unrelated text.

```js
// Unrelated text @jsx h Unrelated text @jsxRuntime classic

/**
  * Unrelated text.
  * @jsx h
  * Unrelated text.
  * @jsxRuntime classic
  */
```

This is more liberal than Babel, which [has a test](https://github.com/babel/babel/blob/226318d54eaff275500b2723c46ee608c0632df0/packages/babel-plugin-transform-react-jsx/test/fixtures/react/should-not-allow-jsx-pragma-to-be-anywhere-in-comment/input.js) specifically to outlaw placing pragmas amongst other text in a comment - this test is disabled in this PR.

Instead, we're aligning with ESBuild ([ESBuild playground](https://esbuild.github.io/try/#dAAwLjI1LjQALS1sb2FkZXI9anN4IC0tanN4PWF1dG9tYXRpYwAvLyBAanN4IFNvbWV0aGluZyBibGFoIGJsYWggQGpzeFJ1bnRpbWUgY2xhc3NpYwo8YmxhaC8+Cg)) which seems to allow a `@jsx` pragma anywhere in the text of the comment.
